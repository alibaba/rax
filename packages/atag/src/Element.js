import { PolymerElement, html } from '@polymer/polymer';

const { HTMLElement } = window;

export { HTMLElement, PolymerElement, html as polymerHtml };
