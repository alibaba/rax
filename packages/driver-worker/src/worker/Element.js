import Node from './Node';
import { ELEMENT_NODE } from './NodeTypes';
import { dispatchAnimationToStyle } from './Animation';
import { mutate } from './MutationObserver';
import camelCase from '../shared/camelCase';
import findWhere from '../shared/findWhere';
import splice from '../shared/splice';
import memorized from '../shared/memorized';

const IS_DATASET_REG = /^data-/;
const elementConstructors = {};

export function registerElement(nodeName, elementConstructor) {
  elementConstructors[nodeName] = elementConstructor;
}

export function createElement(nodeName) {
  return createElementNS(null, nodeName);
}

export function createElementNS(namespaceURI, nodeName) {
  return new (elementConstructors[nodeName] || Element)(ELEMENT_NODE, nodeName, namespaceURI);
}

export function isElement(node) {
  return node && node.nodeType === ELEMENT_NODE;
}

const isDataset = memorized(function(attrName) {
  return IS_DATASET_REG.test(attrName);
});

const createAttributeFilter = memorized(function(ns, name) {
  return o => o.ns === ns && String(o.name).toLowerCase() === String(name).toLowerCase();
});

export default class Element extends Node {
  attributes = [];
  _style = {};

  constructor(nodeType, nodeName, namespaceURI) {
    super(nodeType || ELEMENT_NODE, nodeName.toUpperCase());
    this._namespace = namespaceURI;
  }

  set className(val) {
    this.setAttribute('class', val);
  }

  get className() {
    return this.getAttribute('class');
  }

  set id(val) {
    this.setAttribute('id', val);
  }

  get id() {
    return this.getAttribute('id');
  }

  set animation(queues) {
    this.setAttribute('animation', queues);
    if (Array.isArray(queues) && queues.length > 0) {
      const handleAnimationQueue = () => {
        if (queues.length > 0) {
          dispatchAnimationToStyle(this, queues.shift());
        } else {
          this.removeEventListener('transitionend', handleAnimationQueue);
        }
      };
      if (queues.length > 0) {
        dispatchAnimationToStyle(this, queues.shift());
        this.addEventListener('transitionend', handleAnimationQueue);
      }
    }
  }
  get animation() {
    return this.getAttribute('animation');
  }

  get children() {
    return this.childNodes.filter(isElement);
  }

  get dataset() {
    const dataset = {};
    for (let i = 0, l = this.attributes.length; i < l; i++) {
      const attr = this.attributes[i];
      if (isDataset(attr.name)) {
        // Remove `data-`
        dataset[camelCase(attr.name.slice(5))] = attr.value;
      }
    }
    return dataset;
  }

  get style() {
    return this._style;
  }

  set style(styleObject) {
    this._style = styleObject;
    mutate(this, 'attributes', { style: styleObject });
  }

  setAttribute(key, value) {
    if (value !== this.getAttribute(key)) {
      let attr = findWhere(this.attributes, createAttributeFilter(undefined, key));
      if (!attr) this.attributes.push(attr = { name: key });
      attr.value = value;
      mutate(this, 'attributes', { attributeName: key, newValue: value });
    }
  }

  getAttribute(key) {
    return this.getAttributeNS(undefined, key);
  }

  removeAttribute(key) {
    this.removeAttributeNS(undefined, key);
  }

  setAttributeNS(ns, name, value) {
    let attr = findWhere(this.attributes, createAttributeFilter(ns, name));
    if (!attr) this.attributes.push(attr = { ns, name });

    attr.value = value;
    mutate(this, 'attributes', { attributeName: name, newValue: value });
  }

  getAttributeNS(ns, name) {
    let attr = findWhere(this.attributes, createAttributeFilter(ns, name));
    return attr && attr.value;
  }

  removeAttributeNS(ns, name) {
    splice(this.attributes, createAttributeFilter(ns, name));
    mutate(this, 'attributes', { attributeName: name });
  }
}
