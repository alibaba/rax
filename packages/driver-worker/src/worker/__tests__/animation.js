import { createElement } from '../Element';

describe('Animation', () => {
  function animate(animation) {
    return [{
      config: {
        transformOrigin: 'top right',
        duration: 3000,
        timeFunction: 'ease-in-out',
        delay: 100
      },
      animation,
    }];
  }

  it('merge style', () => {
    const animationConfig = animate([['translate', [35]]]);
    const el = createElement('foo');
    el.style = {
      transform: 'rotate(30deg)'
    };
    el.animation = animationConfig;
    expect(el.animation).toEqual(animationConfig);
    expect(el.style).toMatchSnapshot();
  });

  it('translate', () => {
    const animationConfig = animate([['translate', [35]]]);
    const el = createElement('foo');
    el.animation = animationConfig;
    expect(el.animation).toEqual(animationConfig);
  });

  it('translate3d', () => {
    const animationConfig = animate([['translate3d', [35, 40, 0]]]);
    const el = createElement('foo');
    el.animation = animationConfig;
    expect(el.animation).toEqual(animationConfig);
  });

  it('scale', () => {
    const animationConfig = animate([['scale', [1.5]]]);
    const el = createElement('foo');
    el.animation = animationConfig;
    expect(el.animation).toEqual(animationConfig);
  });

  it('scale3d', () => {
    const animationConfig = animate([['scale3d', [1.5, 1.2, 1.0]]]);
    const el = createElement('foo');
    el.animation = animationConfig;
    expect(el.animation).toEqual(animationConfig);
  });

  it('rotateX', () => {
    const animationConfig = animate([['rotateX', [30]]]);
    const el = createElement('foo');
    el.animation = animationConfig;
    expect(el.animation).toEqual(animationConfig);
  });

  it('rotate3d', () => {
    const animationConfig = animate([['rotate3d', [30, 40, 50]]]);
    const el = createElement('foo');
    el.animation = animationConfig;
    expect(el.animation).toEqual(animationConfig);
  });
});
