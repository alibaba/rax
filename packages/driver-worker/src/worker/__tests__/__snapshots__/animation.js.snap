// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animation merge style 1`] = `
Object {
  "transform": "rotate(30deg)",
}
`;
